File: figcli2/fig-ipc/hook.go
package fig_ipc

import (
	fig_proto "fig-cli/fig-proto"
	"os"
	"strings"
)

const (
	currentIntegrationVersion = 5
)

func SendHook(hook *fig_proto.Hook) error {
	conn, err := Connect()
	if err != nil {
		return err
	}

	message := fig_proto.LocalMessage{
		Type: &fig_proto.LocalMessage_Hook{
			Hook: hook,
		},
	}

	if err = conn.SendFigProto(&message); err != nil {
		return err
	}

	if err = conn.Close(); err != nil {
		return err
	}

	return nil
}

func GenerateShellContext(
	pid int32,
	tty string,
	sessionId string,
	integrationVersion int32,
) (*fig_proto.ShellContext, error) {
	wd, err := os.Getwd()
	if err != nil {
		return nil, err
	}

	shell, err := GetShell()
	if err != nil {
		return nil, err
	}

	return &fig_proto.ShellContext{
		Pid:                     &pid,
		Ttys:                    &tty,
		ProcessName:             &shell,
		CurrentWorkingDirectory: &wd,
		SessionId:               &sessionId,
		IntegrationVersion:      &integrationVersion,
	}, nil
}

func CreateEditBufferHook(sessionId string, integrationVersion int, tty string, pid int, histno int, cursor int, text string) (*fig_proto.Hook, error) {
	context, err := GenerateShellContext(
		int32(pid),
		tty,
		sessionId,
		int32(integrationVersion),
	)

	if err != nil {
		return nil, err
	}

	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_EditBuffer{
			EditBuffer: &fig_proto.EditBufferHook{
				Context: context,
				Text:    text,
				Cursor:  int64(cursor),
				Histno:  int64(histno),
			},
		},
	}, nil
}

func CreatePromptHook(pid int, tty string) (*fig_proto.Hook, error) {
	context, err := GenerateShellContext(
		int32(pid),
		tty,
		os.Getenv("TERM_SESSION_ID"),
		int32(currentIntegrationVersion),
	)

	if err != nil {
		return nil, err
	}

	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_Prompt{
			Prompt: &fig_proto.PromptHook{
				Context: context,
			},
		},
	}, nil
}

func CreateInitHook(pid int, tty string) (*fig_proto.Hook, error) {
	env := os.Environ()
	envMap := make(map[string]string)
	for _, e := range env {
		pair := strings.Split(e, "=")
		envMap[pair[0]] = pair[1]
	}

	term, err := GetCurrentTerminal()
	if err != nil {
		return nil, err
	}

	bundle, err := term.PotentialBundleId()
	if err != nil {
		return nil, err
	}

	context, err := GenerateShellContext(
		int32(pid),
		tty,
		os.Getenv("TERM_SESSION_ID"),
		int32(currentIntegrationVersion),
	)

	if err != nil {
		return nil, err
	}

	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_Init{
			Init: &fig_proto.InitHook{
				Context:      context,
				CalledDirect: false,
				Bundle:       bundle,
				Env:          envMap,
			},
		},
	}, nil
}

func CreateKeyboardFocusChangedHook(bundleIdentifier string, focusedSessionId string) *fig_proto.Hook {
	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_KeyboardFocusChanged{
			KeyboardFocusChanged: &fig_proto.KeyboardFocusChangedHook{
				BundleIdentifier: bundleIdentifier,
				FocusedSessionId: focusedSessionId,
			},
		},
	}
}

func CreateIntegrationReadyHook(identifyier string) *fig_proto.Hook {
	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_IntegrationReady{
			IntegrationReady: &fig_proto.IntegrationReadyHook{
				Identifier: identifyier,
			},
		},
	}
}

func CreateHideHook() *fig_proto.Hook {
	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_Hide{
			Hide: &fig_proto.HideHook{},
		},
	}
}

func CreateEventHook(eventName string) *fig_proto.Hook {
	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_Event{
			Event: &fig_proto.EventHook{
				EventName: eventName,
			},
		},
	}
}

func CreatePreExecHook(pid int, tty string) (*fig_proto.Hook, error) {
	context, err := GenerateShellContext(
		int32(pid),
		tty,
		os.Getenv("TERM_SESSION_ID"),
		int32(currentIntegrationVersion),
	)

	if err != nil {
		return nil, err
	}

	return &fig_proto.Hook{
		Hook: &fig_proto.Hook_PreExec{
			PreExec: &fig_proto.PreExecHook{
				Context: context,
				Command: "",
			},
		},
	}, nil
}

File: dotfilesd/src/ipc/hook.rs
use crate::util::get_shell;
use anyhow::Result;
use fig_proto::local::{
    hook::Hook, EditBufferHook, EventHook, HideHook, InitHook, IntegrationReadyHook,
    KeyboardFocusChangedHook, PreExecHook, PromptHook, ShellContext,
};
use std::collections::HashMap;

const CURRENT_INTEGRATION_VERSION: i32 = 5;

fn generate_shell_context(
    pid: i32,
    tty: impl Into<String>,
    session_id: impl Into<Option<String>>,
    integration_version: Option<i32>,
) -> Result<ShellContext> {
    let cwd = std::env::current_dir()?;
    let shell = get_shell()?;
    Ok(ShellContext {
        pid: Some(pid),
        ttys: Some(tty.into()),
        session_id: session_id
            .into()
            .or_else(|| std::env::var("TERM_SESSION_ID").ok()),
        integration_version: Some(integration_version.unwrap_or(CURRENT_INTEGRATION_VERSION)),
        process_name: Some(shell),
        current_working_directory: Some(cwd.to_string_lossy().into()),
        terminal: None,
        hostname: None,
        remote_context: None,
    })
}

pub fn create_edit_buffer_hook(
    session_id: impl Into<String>,
    integration_version: i32,
    tty: impl Into<String>,
    pid: i32,
    histno: i64,
    cursor: i64,
    text: impl Into<String>,
) -> Result<Hook> {
    let context =
        generate_shell_context(pid, tty, Some(session_id.into()), Some(integration_version))?;
    Ok(Hook::EditBuffer(EditBufferHook {
        context: Some(context),
        text: text.into(),
        cursor,
        histno,
    }))
}

pub fn create_prompt_hook(pid: i32, tty: impl Into<String>) -> Result<Hook> {
    let context = generate_shell_context(pid, tty, None, None)?;
    Ok(Hook::Prompt(PromptHook {
        context: Some(context),
    }))
}

pub fn create_init_hook(pid: i32, tty: impl Into<String>) -> Result<Hook> {
    let env_map: HashMap<_, _> = std::env::vars().collect();
    let context = generate_shell_context(pid, tty, None, None)?;
    Ok(Hook::Init(InitHook {
        context: Some(context),
        called_direct: false,
        bundle: "".into(), // GetCurrentTerminal()?.PotentialBundleId()?
        env: env_map,
    }))
}

pub fn create_keyboard_focus_changed_hook(
    app_identifier: impl Into<String>,
    focused_session_id: impl Into<String>,
) -> Result<Hook> {
    Ok(Hook::KeyboardFocusChanged(KeyboardFocusChangedHook {
        app_identifier: app_identifier.into(),
        focused_session_id: focused_session_id.into(),
    }))
}

pub fn create_integration_ready_hook(identifier: impl Into<String>) -> Result<Hook> {
    Ok(Hook::IntegrationReady(IntegrationReadyHook {
        identifier: identifier.into(),
    }))
}

pub fn create_hide_hook() -> Result<Hook> {
    Ok(Hook::Hide(HideHook {}))
}

pub fn create_event_hook(event_name: impl Into<String>) -> Result<Hook> {
    Ok(Hook::Event(EventHook {
        event_name: event_name.into(),
    }))
}

pub fn create_preexec_hook(pid: i32, tty: impl Into<String>) -> Result<Hook> {
    let context = generate_shell_context(pid, tty, None, None)?;
    Ok(Hook::PreExec(PreExecHook {
        context: Some(context),
        command: None,
    }))
}
